File: Assets/Scripts/EventManager.cs
using System;
using System.Collections.Generic;

namespace LiteMore
{
    public abstract class EventBase
    {
    }

    public static class EventManager
    {
        private abstract class EventListener
        {
            public abstract void Trigger(EventBase Msg);
        }

        private class EventListenerImpl<T> : EventListener where T : EventBase
        {
            public event Action<T> OnEvent = null;

            public override void Trigger(EventBase Msg)
            {
                OnEvent?.Invoke((T)Msg);
            }
        }

        private static readonly Dictionary<string, EventListener> EventList_ = new Dictionary<string, EventListener>();

        public static bool Startup()
        {
            EventList_.Clear();
            return true;
        }

        public static void Shutdown()
        {
        }

        public static void Send<T>(T Event) where T : EventBase
        {
            var EventName = typeof(T).FullName;
            if (EventList_.ContainsKey(EventName))
            {
                ((EventListenerImpl<T>)EventList_[EventName]).Trigger(Event);
            }
        }

        public static void Send<T>() where T : EventBase, new()
        {
            var Event = new T();
            Send(Event);
        }

        public static void Register<T>(Action<T> Callback) where T : EventBase
        {
            var EventName = typeof(T).FullName;
            if (!EventList_.ContainsKey(EventName))
            {
                EventList_.Add(EventName, new EventListenerImpl<T>());
            }

            ((EventListenerImpl<T>)EventList_[EventName]).OnEvent += Callback;
        }

        public static void UnRegister<T>(Action<T> Callback) where T : EventBase
        {
            var EventName = typeof(T).FullName;
            if (EventList_.ContainsKey(EventName))
            {
                ((EventListenerImpl<T>)EventList_[EventName]).OnEvent -= Callback;
            }
        }
    }
}

File: Assets/StandaloneAssets/scripts/framework/manager/eventmanager.lua
EventManager = {}

local eventHandlerList = {}

function EventManager:Startup()
	eventHandlerList = {}
end

function EventManager:Shutdown()
	for k, v in pairs(eventHandlerList) do
        if table.len(v) ~= 0 then
		    print('EventManager : not unregister [' .. k .. ']')
        end
	end
	
	eventHandlerList = {}
end

function EventManager:Register(eventName, eventHandler)
	if type(eventName) ~= 'string' or type(eventHandler) ~= 'function' then
		return
	end

	if eventHandlerList[eventName] == nil then
		eventHandlerList[eventName] = {}
	end

	for k, v in pairs(eventHandlerList[eventName]) do
		if v == eventHandler then
			return
		end
	end

	table.insert(eventHandlerList[eventName], eventHandler)
end

function EventManager:UnRegister(eventName, eventHandler)
	if type(eventName) ~= 'string' or type(eventHandler) ~= 'function' then
		return
	end

	if eventHandlerList[eventName] == nil then
		return
	end

	for k, v in pairs(eventHandlerList[eventName]) do
		if eventHandler == v then
			table.remove(eventHandlerList[eventName], k)
			return
		end
	end
end

function EventManager:Send(event)
	if type (event.name) ~= 'string' then
		return
	end

	if eventHandlerList[event.name] == nil then
		return
	end

	for k, v in pairs(eventHandlerList[event.name]) do
		v(event)
	end
end